php_phongo.c

```diff
@@ -633,11 +737,44 @@ int phongo_execute_command(mongoc_client_t *client, const char *db, zval *zcomma
 				phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
 			}
 
-			bson_free(opts);
 			return false;
 		}
 	}
 
+	return selected_server_id;
+}
+int phongo_execute_command(mongoc_client_t *client, const char *db, zval *zcommand, zval *options, int server_id, zval *return_value, int return_value_used TSRMLS_DC) /* {{{ */
+{
+	const php_phongo_command_t *command;
+	bson_iter_t iter;
+	bson_t reply;
+	bson_error_t error;
+	bson_t *opts;
+	mongoc_cursor_t *cmd_cursor;
+	uint32_t selected_server_id;
+	zval                     *zreadPreference = NULL;
+
+	command = Z_COMMAND_OBJ_P(zcommand);
+
+	opts = bson_new();
+
+	/* FIXME: Legacy way of specifying the readPreference option into this function */
+	if (options && Z_TYPE_P(options) == IS_OBJECT && instanceof_function(Z_OBJCE_P(options), php_phongo_readpreference_ce TSRMLS_CC)) {
+		zreadPreference = options;
+	} else if (!phongo_execute_parse_options(options, PHONGO_READPREFERENCE_ALLOWED, &zreadPreference, NULL TSRMLS_CC)) {
+		return false;
+	}
+
+	selected_server_id = phongo_do_select_server(client, opts, zreadPreference, server_id TSRMLS_CC);
+	if (!selected_server_id) {
+		bson_free(opts);
+		return false;
+	}
+
+	/* As "opts" already always includes the serverId option, the read
+	 * preference isn't actually used here again. The docs for
+	 * mongoc_client_command_with_opts don't actually include the
+	 * read_preference either? */
 	if (!mongoc_client_command_with_opts(client, db, command->bson, phongo_read_preference_from_zval(zreadPreference TSRMLS_CC), opts, &reply, &error)) {
```

Comment:
It's good that we still pass the read prefs here as they do get loaded into the command parts (relevant for mongos command construction) even if we're providing a `serverId` in opts.